Port coroutines to EMTERPRETIFY_ASYNC

atrosinenko opened the pull request:
This code is merely a rewritten coroutine support code from `ASYNCIFY`. It definitely needs review, but it passes the test for coroutine implementation `test_coroutine` (I have splitted it into two tests: one for `ASYNCIFY` and one for `EMTERPRETER_ASYNC`, respectively) and it even seems to work in real project.

atrosinenko commented:
Meanwhile, it looks like coroutine support functions have incorrect asm signatures specified and coroutine allocation function from `ASYNCIFY` [allocates 32 bytes less than needed](https://github.com/atrosinenko/emscripten/blob/b8e4f8c3cfe1e9b043bd4213ad9a960eed4fcd7e/src/library_async.js#L134) because it needs a 32 byte header as well as coroutine stack.

kripken commented on src/library_async.js:
```diff
@@ -410,6 +410,101 @@ mergeInto(LibraryManager.library, {
+
+    // switch context
+    {{{ makeSetValueAsm('coroutine', 0, 'EMTSTACKTOP', 'i32') }}};
+    temp = Module['asm'].emtStackSave();
```
these are unsafe - using `asm` is not supported. instead, those should be exported, and usable as `Module['emtStackSave']()` etc.

kripken replied:
(In other words, `Module['asm']` is an internal detail, that should not be relied on, as little as possible.)

atrosinenko replied:
I have rewritten calls in my PR, but `EmterpreterAsync.handle()` does already contain such calls -- should they be converted too? And what about allocation size in the coroutine creation function from ASYNCIFY and function signatures?

kripken replied:
Yes, if there are other things like that, we should convert them as well.

Not sure what you mean about allocation size? If you mean that asyncify also does that, then yes, the same principle applies there, we should use things on `Module`, avoiding `Module['asm']`.

atrosinenko replied:
I mean shouldn't we [here](https://github.com/atrosinenko/emscripten/blob/b8e4f8c3cfe1e9b043bd4213ad9a960eed4fcd7e/src/library_async.js#L134) use `coroutine = _malloc((stack_size + 32) | 0)|0;` instead?

kripken replied:
I see, thanks. Yes, you're right, the allocation should add the stack plus the extra space we need.

kripken commented on src/library_async.js:
```diff
@@ -410,6 +410,101 @@ mergeInto(LibraryManager.library, {
+  /*
+   * Layout of an EMTERPRETIFY_ASYNC coroutine structure:
+   *
+   *  0 callee's EMTSTACKTOP
```
I don't think we need to track the global var `EMTSTACKTOP` etc. - we do set it up in JS, but then responsibility is in the compiled code. So we just need that.

atrosinenko replied:
`EMTSTACKTOP` is used at least [here](https://github.com/atrosinenko/emscripten/blob/b8e4f8c3cfe1e9b043bd4213ad9a960eed4fcd7e/src/library_async.js#L265). And so [here](https://github.com/atrosinenko/emscripten/blob/b8e4f8c3cfe1e9b043bd4213ad9a960eed4fcd7e/src/library_async.js#L309) it is used to find the first stack frame (I use it likewise in my PR) -- as I understand, it is the global var, not one from the compiled code. Meanwhile, if it is used only for initialization, it would be great to rename it like `INITIAL_EMTSTACKTOP` -- it took some time for me to understand why one creates subarray from `EMTSTACKTOP` to `emtStackSave()` that returns `EMTSTACKTOP` too. :) But it is another story, of course.

kripken replied:
I see, thanks. Looks like it's used as the very initial start of the stack, yeah. During startup, we should copy it to `EMTSTACK_BASE` and use that (we do a similar thing for the normal stack), and otherwise never use `EMTSTACKTOP` in JS.

atrosinenko commented:
By suspicious signatures I mean: shouldn't `emscripten_coroutine_create__sig` be `iiii` and not `iii`: it takes 3 ints and return int?

kripken commented:
Yes, that signature should be `iiii`, you are right. I guess we didn't actually need it (it's used in function tables and a few other things, but not all the time). Yes, we should fix it.

atrosinenko commented:
I have converted other uses of `Module['asm']` in `library_async.js` supposing `stackSave` is exported. `emscripten.py` contains the following code:
```python
    if not settings['SIDE_MODULE']:
      funcs += ['stackAlloc', 'stackSave', 'stackRestore', 'establishStackSpace']
```
(and all of that is under `  if not settings['ONLY_MY_CODE']:`).
Meanwhile, I don't save non-emterpreter stack in my coroutine code (`stackSave` / `stackRestore`) -- is it OK?

And is it correct to reference exported function through `Module['function']` without referencing them in `__deps`?

atrosinenko commented:
Now already listed issues are fixed. I didn't touched `EMTSTACK_BASE` at all for now, though.

kripken commented:
Thanks! Everything looks great, except there is a merge conflict to resolve, and for some reason `binaryen*.test_complex` is failing, but I'm sure that's not the fault of this PR, I'm investigating that now.

kripken commented:
I can't reproduce those failures, very odd. When you resolve the merge conflict, hopefully they'll go away.

atrosinenko commented:
I have rebased my PR against new incoming and added a TODO on `EMTSTACK_BASE`.

Meanwhile, the `test_coroutine_emterpretify_async` takes about 6.5s to complete, while `test_coroutine_asyncify` takes only 1.5s -- hope it's OK. And newly added `test_async_abort` is marked as emterpreter-only. Should I mark my test likewise?

kripken commented:
Looks good, thanks!

In the future, I think it's usually easier to merge than to rebase in PRs. The merge itself is easier, and then the new diff in the PR is also clearer. Before github had the squash option it was better to rebase, but now if you squash it's the same anyhow, so merging is fine.

Odd that test takes 6 seconds - the emterpreter is slow, but I didn't think that test did that much work. Might be worth shortening it later.

No need to mark the new tests here, we do want them to run in other modes (and we enable emterpreter in them, that way we test more things).

atrosinenko commented:
> Odd that test takes 6 seconds - the emterpreter is slow, but I didn't think that test did that much work. Might be worth shortening it later.

It looks like while this test run in about 6.2s on my machine, the `/tmp/emscripten_temp/src.cpp.o.js`left from this test (with `EM_SAVE_DIR=1` set) runs in just about 220ms -- not too short, but far less than the total test run time.

kripken commented:
Yes, testing this, most of the time is compilation. The emterpreter compilation process is pretty slow...
